src/testing/run.ts

```diff
@@ -0,0 +1,282 @@
+import { BaseTestEvaluator, type Evaluation } from './models';
+import { AutoblocksEnvVar, readEnv } from '../util';
+import crypto from 'crypto';
+
+const DEFAULT_MAX_TEST_CASE_CONCURRENCY = 10;
+const DEFAULT_MAX_EVALUATOR_CONCURRENCY = 5;
+
+const client = {
+  post: async (path: string, data: unknown): Promise<void> => {
+    const serverAddress = readEnv(
+      AutoblocksEnvVar.AUTOBLOCKS_CLI_SERVER_ADDRESS,
+    );
+    if (!serverAddress) {
+      throw new Error(
+        `\n
+Autoblocks tests must be run within the context of the testing CLI.
+Make sure you are running your test command with:
+$ npx autoblocks testing exec -- <your test command>
+`,
+      );
+    }
+
+    try {
+      await fetch(serverAddress + path, {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify(data),
+      });
+    } catch (err) {
+      // Ignore, any errors for these requests are displayed by the CLI server
+    }
+  },
+};
+
+function isPrimitive(value: unknown): boolean {
+  return (
+    typeof value === 'string' ||
+    typeof value === 'number' ||
+    typeof value === 'boolean'
+  );
+}
+
+function makeTestCaseHash<TestCaseType>(
+  testCase: TestCaseType,
+  testCaseHash:
+    | (keyof TestCaseType & string)[]
```

Comment:
Do we need & string here? Does that break the keyof type?

Will this work with nested properties?

Reply:
yeah otherwise TS will complain that it can be number or symbol, and no doesn't work with nested properties. would need to do something fancier for that

Reply:
got it, thats fine for now